File: go/buildpy/internal/models/dependency.go
package models

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sync"

	"github.com/charmbracelet/log"
	"github.com/shakfu/buildpy/internal/shell"
)

type BuildSystem int16

const (
	CMAKE BuildSystem = iota
	CONFIG
	MAKE
)

type Dependency struct {
	Name        string
	Version     string
	DownloadUrl string
	RepoUrl     string
	RepoBranch  string
	StaticLibs  []string
	Project     *Project
	BuildSys    BuildSystem
}

func (d *Dependency) Prefix() string {
	return filepath.Join(d.Project.Install, d.Name)
}

func (d *Dependency) SrcDir() string {
	return filepath.Join(d.Project.Src, d.Name)
}

func (d *Dependency) BuildDir() string {
	return filepath.Join(d.SrcDir(), "build")
}

func (d *Dependency) StaticLibsExist() bool {
	for _, path := range d.StaticLibs {
		var libpath = filepath.Join(d.Prefix(), "lib", path)
		if _, err := os.Stat(libpath); errors.Is(err, os.ErrNotExist) {
			return false
		}
	}
	return true
}

func (d *Dependency) GitClone() {
	shell.GitClone(d.RepoUrl, d.RepoBranch, d.SrcDir(), false)
}

func InstallOpenssl() {
	ssl := Dependency{
		Name:        "openssl",
		Version:     "1.1.1w",
		DownloadUrl: "https://www.openssl.org/source/old/1.1.1/openssl-1.1.1w.tar.gz",
		RepoUrl:     "https://github.com/openssl/openssl.git",
		RepoBranch:  "OpenSSL_1_1_1w",
		StaticLibs:  []string{"libssl.a", "libcrypto.a"},
		Project:     NewProject(),
		BuildSys:    CONFIG,
	}
	if !ssl.StaticLibsExist() {
		ssl.Project.Setup()
		prefixOpt := fmt.Sprintf("--prefix=%s", ssl.Prefix())
		ssl.GitClone()
		shell.ShellCmd(ssl.SrcDir(), "./config", "no-shared", "no-tests", prefixOpt)
		shell.Make(ssl.SrcDir(), "install_sw")
		if !ssl.StaticLibsExist() {
			log.Fatal("could not build openssl")
		}
	}
}

func InstallOpensslAsync(wg *sync.WaitGroup) {
	defer wg.Done()
	InstallOpenssl()
}

func InstallBzip2() {
	bz2 := Dependency{
		Name:        "bzip2",
		Version:     "1.0.8",
		DownloadUrl: "https://sourceware.org/pub/bzip2/bzip2-1.0.8.tar.gz",
		RepoUrl:     "https://github.com/libarchive/bzip2.git",
		RepoBranch:  "bzip2-1.0.8",
		StaticLibs:  []string{"libbz2.a"},
		Project:     NewProject(),
		BuildSys:    MAKE,
	}
	if !bz2.StaticLibsExist() {
		bz2.Project.Setup()
		prefixOpt := fmt.Sprintf("PREFIX=%s", bz2.Prefix())
		bz2.GitClone()
		shell.Make(bz2.SrcDir(), "install", prefixOpt, "CFLAGS='-fPIC'")
		if !bz2.StaticLibsExist() {
			log.Fatal("could not build bzip2")
		}
	}
}

func InstallBzip2Async(wg *sync.WaitGroup) {
	defer wg.Done()
	InstallBzip2()
}

func InstallXz() {
	xz := Dependency{
		Name:        "xz",
		Version:     "5.6.3",
		DownloadUrl: "https://github.com/tukaani-project/xz/releases/download/v5.6.3/xz-5.6.3.tar.gz",
		RepoUrl:     "https://github.com/python/cpython-source-deps.git",
		RepoBranch:  "xz",
		StaticLibs:  []string{"liblzma.a"},
		Project:     NewProject(),
		BuildSys:    CMAKE,
	}
	if !xz.StaticLibsExist() {
		xz.Project.Setup()
		xz.GitClone()
		os.Chmod(filepath.Join(xz.SrcDir(), "configure"), 0755)
		os.Chmod(filepath.Join(xz.SrcDir(), "build-aux/install-sh"), 0755)
		prefixOpt := fmt.Sprintf("--prefix=%s", xz.Prefix())
		shell.ShellCmd(xz.SrcDir(), "./configure", 
            "--disable-dependency-tracking",
            "--disable-xzdec",
            "--disable-lzmadec",
            "--disable-nls",
            "--enable-small",
            "--disable-shared",
			prefixOpt,
		)
		shell.Make(xz.SrcDir(), "install")
		if !xz.StaticLibsExist() {
			log.Fatal("could not build lzma")
		}
	}

	// if !xz.StaticLibsExist() {
	// 	xz.Project.Setup()
	// 	xz.GitClone()
	// 	var envars = []string{"CFLAGS=-fPIC"}
	// 	shell.CmakeConfigureEnv(xz.SrcDir(), xz.BuildDir(), envars,
	// 		"-DBUILD_SHARED_LIBS=OFF", "-DENABLE_NLS=OFF", "-DENABLE_SMALL=ON",
	// 		"-DCMAKE_BUILD_TYPE=MinSizeRel",
	// 	)
	// 	shell.CmakeBuildEnv(xz.BuildDir(), true, envars)
	// 	shell.CmakeInstall(xz.BuildDir(), xz.Prefix())
	// 	if !xz.StaticLibsExist() {
	// 		log.Fatal("could not build lzma")
	// 	}
	// }
}

func InstallXzAsync(wg *sync.WaitGroup) {
	defer wg.Done()
	InstallXz()
}

File: ml/buildpy/lib/dependency.ml
open Shell

type dependency = {
  name: string;
  version: string;
  download_url: string;
  repo_url: string;
  repo_branch: string option;
  static_libs: string list;
  project: Project.project;
  build_sys: Config.build_system;
}

let create_dependency name version download_url repo_url ?repo_branch static_libs project build_sys =
  { name; version; download_url; repo_url; repo_branch; static_libs; project; build_sys }

let prefix dep = Filename.concat dep.project.install dep.name

let src_dir dep = Filename.concat dep.project.src dep.name

let build_dir dep = Filename.concat (src_dir dep) "build"

let static_libs_exist dep =
  List.for_all (fun lib ->
    let lib_path = Filename.concat (Filename.concat (prefix dep) "lib") lib in
    Sys.file_exists lib_path
  ) dep.static_libs

let git_clone dep =
  let dest = src_dir dep in
  Shell.git_clone ?branch:dep.repo_branch dep.repo_url dest

let install_openssl () =
  let project = Project.create_project () in
  let ssl = create_dependency
    "openssl"
    "1.1.1w"
    "https://www.openssl.org/source/old/1.1.1/openssl-1.1.1w.tar.gz"
    "https://github.com/openssl/openssl.git"
    ~repo_branch:"OpenSSL_1_1_1w"
    ["libssl.a"; "libcrypto.a"]
    project
    Config.CONFIG
  in
  if not (static_libs_exist ssl) then (
    Project.setup ssl.project;
    let prefix_opt = "--prefix=" ^ (prefix ssl) in
    git_clone ssl;
    run_command ~cwd:(Some (src_dir ssl)) "./config" ["no-shared"; "no-tests"; prefix_opt];
    Shell.make ~cwd:(Some (src_dir ssl)) "install_sw" [];
    if not (static_libs_exist ssl) then
      failwith "Could not build openssl"
  )

let install_bzip2 () =
  let project = Project.create_project () in
  let bz2 = create_dependency
    "bzip2"
    "1.0.8"
    "https://sourceware.org/pub/bzip2/bzip2-1.0.8.tar.gz"
    "https://github.com/libarchive/bzip2.git"
    ~repo_branch:"bzip2-1.0.8"
    ["libbz2.a"]
    project
    Config.MAKE
  in
  if not (static_libs_exist bz2) then (
    Project.setup bz2.project;
    let prefix_opt = "PREFIX=" ^ (prefix bz2) in
    git_clone bz2;
    Shell.make ~cwd:(Some (src_dir bz2)) "install" [prefix_opt; "CFLAGS='-fPIC'"];
    if not (static_libs_exist bz2) then
      failwith "Could not build bzip2"
  )

let install_xz () =
  let project = Project.create_project () in
  let xz = create_dependency
    "xz"
    "5.6.3"
    "https://github.com/tukaani-project/xz/releases/download/v5.6.3/xz-5.6.3.tar.gz"
    "https://github.com/python/cpython-source-deps.git"
    ~repo_branch:"xz"
    ["liblzma.a"]
    project
    Config.CMAKE
  in
  if not (static_libs_exist xz) then (
    Project.setup xz.project;
    git_clone xz;
    let configure_script = Filename.concat (src_dir xz) "configure" in
    let install_sh = Filename.concat (Filename.concat (src_dir xz) "build-aux") "install-sh" in
    Unix.chmod configure_script 0o755;
    Unix.chmod install_sh 0o755;
    let prefix_opt = "--prefix=" ^ (prefix xz) in
    let configure_args = [
      "./configure";
      "--disable-dependency-tracking";
      "--disable-xzdec";
      "--disable-lzmadec";
      "--disable-nls";
      "--enable-small";
      "--disable-shared";
      prefix_opt;
    ] in
    run_command ~cwd:(Some (src_dir xz)) (List.hd configure_args) (List.tl configure_args);
    Shell.make ~cwd:(Some (src_dir xz)) "install" [];
    if not (static_libs_exist xz) then
      failwith "Could not build lzma"
  )

let install_dependencies () =
  log_info "Installing dependencies...";
  install_openssl ();
  install_bzip2 ();
  install_xz ()